lib/parser.ex

```diff
@@ -0,0 +1,205 @@
+defmodule Membrane.Ogg.Parser do
+  import Bitwise
+
+  @type packet :: %{payload: binary, track_id: integer, bos?: boolean, eos?: boolean}
+
+  @spec parse(binary, continued_packets :: %{}) ::
+          {parsed :: list(packet), unparsed :: binary, continued_packets :: %{}}
+  def parse(bytes, continued_packets) do
+    do_parse([], bytes, continued_packets)
+  end
+
+  @spec do_parse(acc :: list(packet), bytes :: binary, continued_packets :: %{}) ::
+          {parsed :: list(packet), bytes :: binary, continued_packets :: %{}}
+  defp do_parse(acc, bytes, continued_packets) do
+    case maybe_parse_page(bytes, continued_packets) do
+      {:need_more_bytes} ->
+        {acc, bytes, continued_packets}
+
+      {:ok, unparsed, packets, continued_packets} ->
+        do_parse(acc ++ packets, unparsed, continued_packets)
+    end
+  end
+
+  @spec maybe_parse_page(page :: binary, continued_packets :: %{}) ::
+          {:need_more_bytes}
+          | {:ok, bytes :: binary, packets :: list(packet), continued_packets :: %{}}
```

Comment:
The convention is to use `{:ok, ...} | {:error, reason}`. If you don't need to return anything together with `:ok` you can write `:ok | {:error, reason}`

```suggestion
          {:ok, bytes :: binary, packets :: list(packet), continued_packets :: %{}}
          | {:error, :need_more_bytes}
```

However, we don't require `spec` for private functions so if you want to you can omit them